cluster-autoscaler/cloudprovider/gce/gce_manager.go

```diff
@@ -809,62 +819,169 @@ func (m *gceManagerImpl) getTemplates() *templateBuilder {
-		if err != nil {
+		glog.V(2).Infof("Refreshed MIG list, next refresh after %v", m.lastRefresh.Add(refreshInterval))
+	case ModeGKENAP:
+		if err := m.fetchResourceLimiter(); err != nil {
+			glog.Errorf("Failed to fetch resource limits: %v", err)
 			return err
 		}
-
-		m.lastRefresh = time.Now()
-		glog.V(2).Infof("Refreshed NodePools list and resource limits, next refresh after %v", m.lastRefresh.Add(refreshInterval))
-		return nil
+		glog.V(2).Infof("Refreshed resource limits, next refresh after %v", m.lastRefresh.Add(refreshInterval))
+		fallthrough
+	case ModeGKE:
+		if err := m.fetchAllNodePools(); err != nil {
+			glog.Errorf("Failed to fetch node pools: %v", err)
+			return err
+		}
+		glog.V(2).Infof("Refreshed NodePool list, next refresh after %v", m.lastRefresh.Add(refreshInterval))
 	}
 	return nil
 }
 
-func (m *gceManagerImpl) fetchResourceLimiter() error {
-	if m.mode == ModeGKENAP {
-		cluster, err := m.gkeAlphaService.Projects.Zones.Clusters.Get(m.projectId, m.location, m.clusterName).Do()
+// Fetch explicitly configured MIGs. These MIGs should never be unregistered
+// during refreshes, even if they no longer exist in GCE.
+func (m *gceManagerImpl) fetchExplicitMigs(specs []string) error {
+	m.assertGCE()
+
+	changed := false
+	for _, spec := range specs {
+		mig, err := m.buildMigFromSpec(spec)
 		if err != nil {
 			return err
 		}
-		if cluster.Autoscaling == nil {
-			return nil
+		if m.RegisterMig(mig) {
+			changed = true
+		}
+		m.neverUnregister[mig.GceRef] = true
+	}
+
+	if changed {
+		if err := m.regenerateCache(); err != nil {
+			return err
 		}
+	}
+	return nil
+}
+
+func (m *gceManagerImpl) buildMigFromSpec(spec string) (*Mig, error) {
+	s, err := dynamic.SpecFromString(spec, scaleToZeroSupported)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse node group spec: %v", err)
+	}
+	mig := &Mig{gceManager: m, minSize: s.MinSize, maxSize: s.MaxSize, exist: true}
+	if mig.Project, mig.Zone, mig.Name, err = ParseMigUrl(s.Name); err != nil {
+		return nil, fmt.Errorf("failed to parse mig url: %s got error: %v", s.Name, err)
+	}
+	return mig, nil
+}
 
-		minLimits := make(map[string]int64)
-		maxLimits := make(map[string]int64)
-		for _, limit := range cluster.Autoscaling.ResourceLimits {
-			if _, found := supportedResources[limit.Name]; !found {
-				glog.Warning("Unsupported limit defined %s: %d - %d", limit.Name, limit.Minimum, limit.Maximum)
+// Fetch automatically discovered MIGs. These MIGs should be unregistered if
+// they no longer exist in GCE.
+func (m *gceManagerImpl) fetchAutoMigs() error {
+	m.assertGCE()
+
+	exists := make(map[GceRef]bool)
+	changed := false
+	for _, cfg := range m.autoMIGs {
+		links, err := m.findMigsNamed(cfg.Re)
+		if err != nil {
+			return fmt.Errorf("cannot autodiscover managed instance groups: %s", err)
+		}
+		for _, link := range links {
+			mig, err := m.buildMigFromAutoCfg(link, cfg)
+			if err != nil {
+				return err
+			}
+			if m.RegisterMig(mig) {
```

Comment:
Good point - this would override the min/max for an explicit MIG. It's reasonable that explicit MIGs would override the min/max of autodiscovered ones. I'll need to fix this here and in the AWS manager.